tools/specs-checker/check.go

```diff
@@ -0,0 +1,176 @@
+	return nil
+}
+
+// parseSpecs parses input spec docs into map of function name -> array of function bodies
+// (single entity may have several definitions).
+func parseSpecs() (map[string][]string, error) {
+	loadSpecsFile := func(sb *strings.Builder, specFilePath string) error {
+		chunk, err := specFS.ReadFile(specFilePath)
+		if err != nil {
+			return fmt.Errorf("cannot read specs file: %w", err)
+		}
+		_, err = sb.Write(chunk)
+		if err != nil {
+			return fmt.Errorf("cannot copy specs file: %w", err)
+		}
+		return nil
+	}
+
+	// Traverse all spec files, and aggregate them within as single string.
+	var sb strings.Builder
+	for dirName, fileNames := range specDirs {
+		for _, fileName := range fileNames {
+			if err := loadSpecsFile(&sb, path.Join("data", dirName, fileName)); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// Load file with extra definitions (this allows us to use pseudo-code that is not from specs).
+	if err := loadSpecsFile(&sb, path.Join("data", "extra.md")); err != nil {
+		return nil, err
+	}
+
+	// Parse docs into function name -> array of function bodies map.
+	chunks := strings.Split(strings.ReplaceAll(sb.String(), "```python", ""), "```")
+	defs := make(map[string][]string, len(chunks))
+	for _, chunk := range chunks {
+		defName, defBody := parseDefChunk(chunk)
+		if defName == "" {
+			continue
+		}
+		defs[defName] = append(defs[defName], defBody)
+	}
+	return defs, nil
+}
+
+// parseDefChunk extract function name and function body from a Python's "def" chunk.
+func parseDefChunk(chunk string) (string, string) {
+	chunk = strings.TrimLeft(chunk, "\n")
+	if chunk == "" {
+		return "", ""
+	}
+	chunkLines := strings.Split(chunk, "\n")
+	// Ignore all snippets, that do not define functions.
+	if chunkLines[0][:4] != "def " {
+		return "", ""
+	}
+	defMatches := reg1.FindStringSubmatch(chunkLines[0])
+	if len(defMatches) < 2 {
+		return "", ""
+	}
+	return strings.Trim(defMatches[1], " "), chunk
+}
+
+// matchesRefImplementation compares input string to reference code snippets (there might be multiple implementations).
+func matchesRefImplementation(defName string, refDefs []string, input string, pos token.Position) bool {
+	for _, refDef := range refDefs {
+		refDefLines := strings.Split(strings.TrimRight(refDef, "\n"), "\n")
+		inputLines := strings.Split(strings.TrimRight(input, "\n"), "\n")
+
+		matchesPerfectly := true
+		for i := 0; i < len(refDefLines); i++ {
+			a, b := strings.Trim(refDefLines[i], " "), strings.Trim(inputLines[i], " ")
+			if a != b {
+				matchesPerfectly = false
+				break
+			}
+		}
+		// Mark potential issues, when there's some more comments in our code (which might be ok, as we are not required
+		// to put specs comments as the last one in the doc block).
+		if checkNumRows && len(refDefLines) != len(inputLines) {
+			fmt.Printf("%s: %q potentially has issues (comment is longer than reference implementation)\n", pos, defName)
+		}
+		if matchesPerfectly {
+			return true
+		}
```

Comment:
Do you want to check that all refDefs match perfectly or just any of them match perfectly?

Reply:
I loop through all versions of a given definition (say, in `altair` we have the function with the same name but different implementation, I want to make sure that both comments in altair-related code and phase0 related code are checked).

So, at the beginning of loop that checks that all definition's lines match our comment, the `matchesPerfectly` flag is set to `true`. Line checking loop breaks when non-matching line is found, the `matchesPerfectly` flag is set to `false`. So, a single matching definition is enough for flag to survive, and us be sure that our comment matches at least one known version of spec's function.

NB: Will search rules go for an issue, and if none exists will create my own -- somehow missed that part, thanks for pointing that out.

Reply:
> This is amazing. Is there any issue filed with rules_go to support embed in go tool library?

- I've created an issue: https://github.com/bazelbuild/rules_go/issues/2860
- Btw, while looking whether the issue has already been filed, I found an alternative way to implement embedding (`go_embed_data` is mentioned in https://github.com/bazelbuild/rules_go/issues/2775). So, it seems we can implement the checker as CI static code analyzer with [go_embed_data](https://github.com/bazelbuild/rules_go/blob/master/go/extras.rst#go-embed-data), do you think it is worth implementing it right now, or should we wait for update on `rules_go` side?